components/renku_data_services/data_connectors/db.py

```diff
@@ -0,0 +1,225 @@
+"""Adapters for data connectors database classes."""
+
+from collections.abc import Callable
+from typing import TypeVar
+
+from sqlalchemy import Select, func, select
+from sqlalchemy.ext.asyncio import AsyncSession
+from ulid import ULID
+
+from renku_data_services import base_models, errors
+from renku_data_services.base_api.pagination import PaginationRequest
+from renku_data_services.data_connectors import apispec, models
+from renku_data_services.data_connectors import orm as schemas
+from renku_data_services.namespace import orm as ns_schemas
+from renku_data_services.utils.core import with_db_transaction
+
+
+class DataConnectorRepository:
+    """Repository for data connectors."""
+
+    def __init__(
+        self,
+        session_maker: Callable[..., AsyncSession],
+    ) -> None:
+        self.session_maker = session_maker
+
+    async def get_data_connectors(
+        self, pagination: PaginationRequest, namespace: str | None = None
+    ) -> tuple[list[models.DataConnector], int]:
+        """Get multiple data connectors from the database."""
+        async with self.session_maker() as session:
+            stmt = select(schemas.DataConnectorORM)
+            if namespace:
+                stmt = _filter_by_namespace_slug(stmt, namespace)
+            stmt = stmt.limit(pagination.per_page).offset(pagination.offset)
+            stmt = stmt.order_by(schemas.DataConnectorORM.id.desc())
+            stmt_count = select(func.count()).select_from(schemas.DataConnectorORM)
+            if namespace:
+                stmt_count = _filter_by_namespace_slug(stmt_count, namespace)
+            results = await session.scalars(stmt), await session.scalar(stmt_count)
```

Comment:
suggestion: [asyncio.gather](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) would allow these to be run concurrently

Reply:
See https://github.com/SwissDataScienceCenter/renku-data-services/issues/404, this would be invalid use of concurrency.